### ISSUE TITLE
Use-after-free reported by clang Address Sanitizer in test-api

### ISSUE THREAD

**Issue by fishilico** [VirusTotal/yara#915]
Hi,
After compiling yara with clang's address sanitizer, `make check` found a heap-use-after-free in `test-api`:
```
==7301==ERROR: AddressSanitizer: heap-use-after-free on address 0xf2bc8819 at pc 0x5660e00a bp 0xffb31468 sp 0xffb31040
READ of size 1 at 0xf2bc8819 thread T0
    #0 0x5660e009 in __interceptor_strlen.part.31 (/usr/src/yara/test-api+0x9a009)
    #1 0x566db3ae in yr_object_from_external_variable /usr/src/yara/libyara/object.c:309:32
    #2 0x566e6455 in yr_scanner_create /usr/src/yara/libyara/scanner.c:187:5
    #3 0x566ac8e1 in test_scanner /usr/src/yara/tests/test-api.c:381:7
    #4 0x566af5d2 in main /usr/src/yara/tests/test-api.c:678:3
    #5 0xf79ea140 in __libc_start_main (/usr/lib32/libc.so.6+0x1a140)
    #6 0x565a14e7 in _start (/usr/src/yara/test-api+0x2d4e7)
```
(The addresses are 32-bit-wide because I was compiling in 32-bit mode).

The faulting address (`0xf2bc8819`) matches the one allocated to `val` in `_yr_compiler_define_variable`: https://github.com/VirusTotal/yara/blob/307ba3d71947dfbf7631c1ddfb83e2334ec9f960/libyara/compiler.c#L909-L919

Adding a relocation with `yr_arena_make_ptr_relocatable` seems to fix this issue:
```c
  if (external->type == EXTERNAL_VARIABLE_TYPE_STRING)
  {
    char* val;

    FAIL_ON_COMPILER_ERROR(yr_arena_write_string(
        compiler->sz_arena,
        external->value.s,
        &val));

    ext->value.s = val;

    FAIL_ON_COMPILER_ERROR(yr_arena_make_ptr_relocatable(
        compiler->externals_arena,
        ext,
        offsetof(YR_EXTERNAL_VARIABLE, value.s),
        EOL));
  }
```

Nevertheless the code looks strange to me, because `ext->value` does not seem to be initialized for non-string types, so I may have missed something here. As this code has been changed 2 days ago (with 
https://github.com/VirusTotal/yara/pull/909), there may be a code refactoring in progress... How should this use-after-free issue be fixed?

**Comment by plusvic** [comment]
Good catch! Your fix was correct and you were also right about the value of non-string external variables not being initialized. This passed unnoticed because users usually initialize external variables by calling `yr_rules_define_xxx_variable`, but the correct behaviour is that `yr_compiler_define_xxx_variable` sets the initial value for the variable that you can later override with `yr_rules_define_xxx_variable`.

I've fixed both issues and added test cases in ef7d5ecb196fdcbed83881d440d127d015d08c5b.

**Comment by reijin90** [comment]
@fishilico If I may ask, how did you compile it? I'm failing to get a working build with:
```
FLAGS="-O1 -g -fsanitize=fuzzer-no-link,address"
CC=~/llvm-install/bin/clang CXX=~/llvm-install/bin/clang++ LD=~/llvm-install/bin/clang CFLAGS="$FLAGS" CXXFLAGS="$FLAGS" ./configure
```
